blueprint: Update to deploy simple website.

kayousterhout opened the pull request:
This commit changes the nginx spec to deploy a very
simple index.html page.

luise commented on main.js:
```diff
@@ -1,17 +1,68 @@
+        '. Run ssh-keygen command to generate one.');
+}
+var sshPublicKey = fs.readFileSync(publicKeyFile, 'utf8');
+console.warn(sshPublicKey);
```
Why is this a warning? And is it that helpful to log the key? If it's just to clarify what's happening under the hood, would it be more helpful to log the path in `publicKeyFile`?

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
-    // Replace with your GitHub username to allow logging into machines.
-    // sshKeys: githubKeys("CHANGE_ME"),
-});
+// Allow inbound connections from port 80. This should be the same as the port
```
I think this is a good enough reason to leave in the templating for the port. Even if we don't let users configure the port, it seems like better code to make it so that there's not this dependency between this constant, and the configuration file.

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
-});
+// Allow inbound connections from port 80. This should be the same as the port
+// that nginx is configured to listen on in the configuration file.
+var port = 80
```
Missing semicolon

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
-// Create Master and Worker Machines.
-deployment.deploy(baseMachine.asMaster())
-deployment.deploy(baseMachine.asWorker().replicate(1));
+var configFilename = 'nginx_defaults.conf'
```
These two lines are missing semicolons as well

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
+// run nginx (one Quilt master and one worker) and returns the resulting
+// deployment.
+function createInfrastructure(publicKey) {
+    var deployment = createDeployment({});
```
No need for the empty object (`{}`) to `createDeployment` anymore (it used to be necessary)

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
+
+// Get a SSH public key to use by looking for one in ~/.ssh/id_rsa.pub.
+var publicKeyFile = path.join(process.env.HOME, '.ssh/id_rsa.pub');
+if (!fs.existsSync(publicKeyFile)) {
```
Should we be requiring that users deploy with a sshKey set? The deployment would still work without one, they just wouldn't be able to use some debug command (`quilt logs`, `quilt ssh`).

kklin replied:
Did you see this comment?

kklin replied:
FWIW, I personally don't have an id_rsa key because I name my keys. Don't know if I'm an exception though.

kayousterhout replied:
In this case I'd like to leave it, because this is the spec in the getting started guide, and I can imagine someone ssh-ing around to make sure that things work.

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
+var publicKeyFile = path.join(process.env.HOME, '.ssh/id_rsa.pub');
+if (!fs.existsSync(publicKeyFile)) {
+    throw new Error('No SSH public key found in ' + publicKeyFile +
+        '. Run ssh-keygen command to generate one.');
```
"Run _the_ ssh-keygen command ..."?

kklin commented on main.js:
```diff
@@ -1,17 +1,68 @@
+    var webTier = new Service('nginx_example', [
+            new Container('nginx:1.13').withFiles({
+                '/etc/nginx/conf.d/default.conf': configFileData,
+                '/site/index.html': indexFileData
```
If you go the templating route, it might also be nice to use templating to fill in the root.

kklin commented:
Oh, if you change this spec, you should search for any importers as well (we may use it in the integration tests)

kayousterhout commented:
I (finally) addressed the comments on this.  Thanks for reviewing!  The new version of the getting started guide uses this new version of nginx as the example.

luise commented on app.js:
```diff
@@ -1,36 +1,57 @@
+ * @return {Service} A service that wraps the container running nginx.
+ */
+exports.createService = function(port) {
+  port = port || 80;
```
Would be better to use a default argument -- i.e. `function(port = 80)`

luise commented on app.js:
```diff
@@ -1,36 +1,57 @@
-    publicInternet.connect(port, webTier);
+  const nginxConfTemplate = fs.readFileSync(
+      path.join(__dirname, 'nginx_defaults.conf.tmpl'), {encoding: 'utf8'});
+  const nginxConf = applyTemplate(
```
Have you looked into using a templating module like [Mustache](https://www.npmjs.com/package/mustache)?

kayousterhout replied:
Also FYI I added this to the Trello card: https://trello.com/c/q8ZoHw7I/575-add-templating-to-simplify-passing-configuration-information-to-containers

kklin replied:
How does the `index.html` actually get placed into the container? I was expecting it to be part of `withFiles`

luise commented on main.js:
```diff
@@ -1,17 +1,32 @@
+}
+const sshPublicKey = fs.readFileSync(publicKeyFile, 'utf8');
+
+const deployment = createDeployment({});
```
You can leave out the `{}`

kklin replied:
Just a question for my own understanding -- when should we use `const`? Here, the deployment object isn't a "constant", you're still changing the object itself later via the `deploy` functions, just not reassigning the reference.

Is the correct style to use `const` for variable references that aren't reassigned, even if the reference is a mutable object?

kayousterhout commented:
Thanks for reviewing Luise!  The comments you left were all on code that I didn't change (you may have noticed this -- but I wanted to mention it just in case you didn't, because the linting commit made the diff misleading!).  I fixed two of them but not the templating, because that's something on the roadmap as something to think about more holistically in the long term (and it's a more substantial fix).

luise commented:
Ah, @kayousterhout , thanks, I didn't realize that -- I just looked at all the changes together. Makes sense!

kayousterhout commented:
Re: `const`, I was wondering about this too.  The Google style guide recommends to use const whenever possible: https://google.github.io/styleguide/jsguide.html#features-use-const-and-let

I updated everything else here!
